### ISSUE TITLE
[DeviceManager] Call failed in template filtering

### ISSUE THREAD

**Issue by betefaber** [dojot/dojot#982]
- templates registered:

![Screenshot_2](https://user-images.githubusercontent.com/37310063/54438696-ef870780-4715-11e9-956c-b9516f1004ea.png)

- filtering by "medidor":

![Screenshot_3](https://user-images.githubusercontent.com/37310063/54438705-f44bbb80-4715-11e9-92f3-e3aeb377a239.png)

- log:

> device-manager_1              | [15/03/19 - 14:14:29] |TemplateHandler| DEBUG: Retrieving templates.
> device-manager_1              | [15/03/19 - 14:14:29] |TemplateHandler| DEBUG: Initializing tenant context...
> device-manager_1              | [15/03/19 - 14:14:29] |TemplateHandler| DEBUG: ... tenant context initialized.
> device-manager_1              | [15/03/19 - 14:14:29] |TemplateHandler| DEBUG: Pagination configuration is {'page': 1, 'per_page': 12, 'error_out': False}
> device-manager_1              | [15/03/19 - 14:14:29] |TemplateHandler| DEBUG: Adding label filter to query...
> device-manager_1              | [15/03/19 - 14:14:29] |TemplateHandler| DEBUG: ... filter was added to query.
> device-manager_1              | [15/03/19 - 14:14:29] |TemplateHandler| DEBUG: Sortby filter is DeviceTemplate.label
> device-manager_1              | [15/03/19 - 14:14:29] |TemplateHandler| DEBUG:  Filtering template by [<sqlalchemy.sql.elements.TextClause object at 0x7fc041e222e8>]
> device-manager_1              | [15/03/19 - 14:14:29] |TemplateHandler| DEBUG: Current query: <class 'flask_sqlalchemy.BaseQuery'>
> device-manager_1              | [2019-03-15 14:14:29,355] ERROR in app: Exception on /template [GET]
> device-manager_1              | Traceback (most recent call last):
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
> device-manager_1              |     context)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 507, in do_execute
> device-manager_1              |     cursor.execute(statement, parameters)
> device-manager_1              | psycopg2.ProgrammingError: SELECT DISTINCT ON expressions must match initial ORDER BY expressions
> device-manager_1              | LINE 2: FROM (SELECT DISTINCT ON (templates.id) templates.id AS id, ...
> device-manager_1              |                                   ^
> device-manager_1              |
> device-manager_1              |
> device-manager_1              | The above exception was the direct cause of the following exception:
> device-manager_1              |
> device-manager_1              | Traceback (most recent call last):
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/flask/app.py", line 1982, in wsgi_app
> device-manager_1              |     response = self.full_dispatch_request()
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/flask/app.py", line 1614, in full_dispatch_request
> device-manager_1              |     rv = self.handle_user_exception(e)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/flask/app.py", line 1517, in handle_user_exception
> device-manager_1              |     reraise(exc_type, exc_value, tb)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/flask/_compat.py", line 33, in reraise
> device-manager_1              |     raise value
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/flask/app.py", line 1612, in full_dispatch_request
> device-manager_1              |     rv = self.dispatch_request()
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/flask/app.py", line 1598, in dispatch_request
> device-manager_1              |     return self.view_functions[rule.endpoint](**req.view_args)
> device-manager_1              |   File "/usr/src/app/DeviceManager/TemplateHandler.py", line 339, in flask_get_templates
> device-manager_1              |     result = TemplateHandler.get_templates(request)
> device-manager_1              |   File "/usr/src/app/DeviceManager/TemplateHandler.py", line 106, in get_templates
> device-manager_1              |     page = BaseQuery(page.subquery(), db.session()).paginate(**pagination)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/flask_sqlalchemy/__init__.py", line 491, in paginate
> device-manager_1              |     items = self.limit(per_page).offset((page - 1) * per_page).all()
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2726, in all
> device-manager_1              |     return list(self)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2878, in __iter__
> device-manager_1              |     return self._execute_and_instances(context)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2901, in _execute_and_instances
> device-manager_1              |     result = conn.execute(querycontext.statement, self._params)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 948, in execute
> device-manager_1              |     return meth(self, multiparams, params)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/sql/elements.py", line 269, in _execute_on_connection
> device-manager_1              |     return connection._execute_clauseelement(self, multiparams, params)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1060, in _execute_clauseelement
> device-manager_1              |     compiled_sql, distilled_params
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1200, in _execute_context
> device-manager_1              |     context)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1413, in _handle_dbapi_exception
> device-manager_1              |     exc_info
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
> device-manager_1              |     reraise(type(exception), exception, tb=exc_tb, cause=cause)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 186, in reraise
> device-manager_1              |     raise value.with_traceback(tb)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
> device-manager_1              |     context)
> device-manager_1              |   File "/usr/src/venv/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 507, in do_execute
> device-manager_1              |     cursor.execute(statement, parameters)
> device-manager_1              | sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) SELECT DISTINCT ON expressions must match initial ORDER BY expressions
> device-manager_1              | LINE 2: FROM (SELECT DISTINCT ON (templates.id) templates.id AS id, ...
> device-manager_1              |                                   ^
> device-manager_1              |  [SQL: "SELECT anon_1.id AS anon_1_id, anon_1.label AS anon_1_label, anon_1.created AS anon_1_created, anon_1.updated AS anon_1_updated \nFROM (SELECT DISTINCT ON (templates.id) templates.id AS id, templates.label AS label, templates.created AS created, templates.updated AS updated \nFROM templates LEFT OUTER JOIN attrs ON templates.id = attrs.template_id \nWHERE templates.label like '%%medidor%%' ORDER BY templates.label) AS anon_1 \n LIMIT %(param_1)s OFFSET %(param_2)s"] [parameters: {'param_1': 12, 'param_2': 0}] (Background on this error at: http://sqlalche.me/e/f405)

**Affected Version:** 58.1-20190313

**Comment by giovannicuriel** [comment]
This might've been fixed in dojot/device-manager#98.

**Comment by betefaber** [comment]
Test OK.

![image](https://user-images.githubusercontent.com/37310063/55828914-d9395500-5ae3-11e9-929d-7ce60565a28b.png)

```
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Retrieving templates.
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Initializing tenant context...
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: ... tenant context initialized.
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Pagination configuration is {'page': 1, 'per_page': 6, 'error_out': False}
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Adding label filter to query...
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: ... filter was added to query.
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Sortby filter is DeviceTemplate.label
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG:  Filtering template by [<sqlalchemy.sql.elements.TextClause object at 0x7fd94d372668>]
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Current query: <class 'flask_sqlalchemy.BaseQuery'>
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Adding resulting template to response...
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Template is: medidor de temperatura
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: ... template was added to response.
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Adding resulting template to response...
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Template is: medidor de pressao
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: ... template was added to response.
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Adding resulting template to response...
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Template is: medidor de umidade relativa
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: ... template was added to response.
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Adding resulting template to response...
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Template is: medidor de velocidade
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: ... template was added to response.
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| DEBUG: Full response is {'pagination': {'page': 1, 'total': 1, 'has_next': False, 'next_page': None}, 'templates': [{'label': 'medidor de temperatura', 'id': 1, 'data_attrs': [{'value_type': 'float', 'label': 'temperatura', 'metadata': [{'value_type': 'string', 'label': 'unidade', 'id': 2, 'updated': None, 'static_value': 'Â°C', 'type': 'meta', 'created': '2019-04-09T11:58:04.461346+00:00'}], 'id': 1, 'template_id': '1', 'type': 'dynamic', 'created': '2019-04-09T11:58:04.458948+00:00'}], 'config_attrs': [], 'created': '2019-04-09T11:58:04.457078+00:00', 'attrs': [{'value_type': 'float', 'label': 'temperatura', 'metadata': [{'value_type': 'string', 'label': 'unidade', 'id': 2, 'updated': None, 'static_value': 'Â°C', 'type': 'meta', 'created': '2019-04-09T11:58:04.461346+00:00'}], 'id': 1, 'template_id': '1', 'type': 'dynamic', 'created': '2019-04-09T11:58:04.458948+00:00'}]}, {'label': 'medidor de pressao', 'id': 2, 'data_attrs': [{'value_type': 'string', 'label': 'SerialNumber', 'id': 3, 'template_id': '2', 'static_value': 'undefined', 'type': 'static', 'created': '2019-04-09T11:58:04.497623+00:00'}, {'value_type': 'float', 'label': 'pressao', 'metadata': [{'value_type': 'string', 'label': 'unidade', 'id': 5, 'updated': None, 'static_value': 'mmHg', 'type': 'meta', 'created': '2019-04-09T11:58:04.499992+00:00'}], 'id': 4, 'template_id': '2', 'type': 'dynamic', 'created': '2019-04-09T11:58:04.498733+00:00'}], 'config_attrs': [], 'created': '2019-04-09T11:58:04.496816+00:00', 'attrs': [{'value_type': 'string', 'label': 'SerialNumber', 'id': 3, 'template_id': '2', 'static_value': 'undefined', 'type': 'static', 'created': '2019-04-09T11:58:04.497623+00:00'}, {'value_type': 'float', 'label': 'pressao', 'metadata': [{'value_type': 'string', 'label': 'unidade', 'id': 5, 'updated': None, 'static_value': 'mmHg', 'type': 'meta', 'created': '2019-04-09T11:58:04.499992+00:00'}], 'id': 4, 'template_id': '2', 'type': 'dynamic', 'created': '2019-04-09T11:58:04.498733+00:00'}]}, {'label': 'medidor de umidade relativa', 'id': 3, 'data_attrs': [{'value_type': 'float', 'label': 'umidade', 'metadata': [{'value_type': 'string', 'label': 'unidade', 'id': 7, 'updated': None, 'static_value': '%', 'type': 'meta', 'created': '2019-04-09T11:58:04.532029+00:00'}], 'id': 6, 'template_id': '3', 'type': 'dynamic', 'created': '2019-04-09T11:58:04.530743+00:00'}], 'config_attrs': [], 'created': '2019-04-09T11:58:04.530030+00:00', 'attrs': [{'value_type': 'float', 'label': 'umidade', 'metadata': [{'value_type': 'string', 'label': 'unidade', 'id': 7, 'updated': None, 'static_value': '%', 'type': 'meta', 'created': '2019-04-09T11:58:04.532029+00:00'}], 'id': 6, 'template_id': '3', 'type': 'dynamic', 'created': '2019-04-09T11:58:04.530743+00:00'}]}, {'label': 'medidor de velocidade', 'id': 4, 'data_attrs': [{'value_type': 'float', 'label': 'velocidade', 'metadata': [{'value_type': 'string', 'label': 'unidade', 'id': 9, 'updated': None, 'static_value': 'km/h', 'type': 'meta', 'created': '2019-04-09T11:58:04.562759+00:00'}], 'id': 8, 'template_id': '4', 'type': 'dynamic', 'created': '2019-04-09T11:58:04.562019+00:00'}], 'config_attrs': [], 'created': '2019-04-09T11:58:04.561294+00:00', 'attrs': [{'value_type': 'float', 'label': 'velocidade', 'metadata': [{'value_type': 'string', 'label': 'unidade', 'id': 9, 'updated': None, 'static_value': 'km/h', 'type': 'meta', 'created': '2019-04-09T11:58:04.562759+00:00'}], 'id': 8, 'template_id': '4', 'type': 'dynamic', 'created': '2019-04-09T11:58:04.562019+00:00'}]}]}
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| INFO:  Getting template with id 1
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| INFO:  Getting template with id 2
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| INFO:  Getting template with id 3
device-manager_1              | [09/04/19 - 19:08:04] |TemplateHandler| INFO:  Getting template with id 4
```
